Implement `soft_margin_loss` as dispatch-less composite kernel.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77486
* #77485
* #77484

This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

ysiraichi committed:
Implement `soft_margin_loss` as dispatch-less composite kernel.

This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/composite/soft_margin_loss.h:
```diff
@@ -0,0 +1,55 @@
+    Tensor& output) {
+  // compute inplace variant of: output = at::log(1. + at::exp(-input *
+  // target));
+  OPS::neg_out(output, input);
```
Doesn't this throw a resize warning when the `output` is the correct ahpw, that is, of shape `{}` whenever `reduction != None`? I believe this should be implemented like:
```c++
TORCH_CHECK(input.sizes() == target.sizes(), "...");
at::native::resize_output(output, output_shape);
auto x = reduction == Reduction::None ? output : at::empty_like(input);
```
and then just use `result` at the end via `mean_out` and`sum_out` if `reduction != None`.

ysiraichi replied:
I think you are right.
Maybe this (and your other suggestion) might be better done in another PR.

Lezcano commented on aten/src/ATen/native/composite/soft_margin_loss.h:
```diff
@@ -0,0 +1,55 @@
+  OPS::neg_out(output, input);
+  OPS::mul_(output, target);
+  OPS::exp_(output);
+  OPS::add_(output, 1.);
+  OPS::log_(output);
```
Don't know if it's supported by this new dispatch-less kernels, but it'd be great to use `log1p` here for numerical stability.

lezcano reviewed (commented):
Left a couple comments, although the problems are coming from the previous implementation, so feel free to ignore.

ysiraichi committed:
Update on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Fix lint errors. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/composite/soft_margin_loss.h:
```diff
@@ -0,0 +1,55 @@
+    int64_t reduction) {
+  auto output = OPS::empty({0}, input.options());
+  OPS::soft_margin_loss_out(output, input, target, reduction);
+  return output;
```
should just make this structured lol

ysiraichi replied:
Hmmm. But, then, isn't true for all composites? i.e. what's the cut-point?

ezyang replied:
Structured is only functional via out composites. So anything else is applicable here.

ysiraichi committed:
Update on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Fix lint. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Fix lint. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Fix merge. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Revert 'add_' dispatch table. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Add 'DispatchlessDEV.cpp' to BAZEL builds. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Raise internal assert if invalid reduction. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ezyang commented:
If debugging the test failures on this is going to take a lot of time I'd rather we spend the time on the overhead reduction backend instead

ysiraichi commented:
Ah. It looks like I messed up: thought the enum would implicitly convert there.

-----

@ezyang @bdhirsh 

Anyway, there was an odd XLA CI failure ([here](https://github.com/pytorch/xla/blob/7c99ad0653b2026fa9ce23045067bfb0badd9c82/torch_xla/csrc/data_ops.cpp#L116)). Midway writing the problem with the XLA build, I think I found the solution.

**The Problem:** I think that XLA [`sum.dim_IntList` implementation](https://github.com/pytorch/xla/blob/08ae1044c2a7e314895f9946104cbe399e096515/torch_xla/csrc/aten_xla_type.cpp#L3198-L3204) does not reduce, given empty dimensions.

1. `SoftMarginLoss` is dispatched to `soft_margin_loss_out_compositeexplicitautograd` (dispatch-less composite kernel). 
2. There, we dispatch [`sum`](https://github.com/pytorch/pytorch/blob/fb6749d977e33b5f463c2d0a1b56a939428105e5/aten/src/ATen/native/ReduceOps.cpp#L1076-L1078) (same for `mean`) **directly** to their `CompositeExplicitAutograd` kernel, even though XLA has an implementation for it ([here](https://github.com/pytorch/xla/blob/08ae1044c2a7e314895f9946104cbe399e096515/torch_xla/csrc/aten_xla_type.cpp#L3188-L3196))
3. `sum` calls its overload with an empty list of dimensions (in PyTorch, it would be interpreted as "all of the dimensions")
4. Then, the overload `sum.IntList_dim` is dispatched to the XLA implementation (since it goes through the dispatcher)
5. In the [XLA implementation](https://github.com/pytorch/xla/blob/08ae1044c2a7e314895f9946104cbe399e096515/torch_xla/csrc/aten_xla_type.cpp#L3198-L3204), however, they never transform the empty list into the "list with all dimensions"
6. By the time it gets to [computing its output size](https://github.com/pytorch/xla/blob/3ceb226cdf32496f8d6270a50f80528a4e014d78/torch_xla/csrc/reduction.cpp#L28-L45), it still remains an empty list

**The Solution:** Change how dispatch-less composites decide what function to dispatch to for each dependency: given a dispatch key, make kernels that aren't registered for that go through the dispatcher. i.e. drop number (2), below. **Obs:** Even if this error didn't happen, I still believe that's a better move. That's because someone may register a kernel for that dispatch key, later. So, I would argue that it makes more sense to go through the dispatcher to get to that, theoretically, faster kernel than to dispatch it to a composite one.


https://github.com/pytorch/pytorch/blob/bb2a2f602a3e2944eba85d51b38b3408560abe1a/torchgen/api/dispatchless.py#L59-L68

ysiraichi committed:
Update on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ysiraichi committed:
Fix lint. on "Implement `soft_margin_loss` as dispatch-less composite kernel."


This PR makes `soft_margin_loss` kernel into a dispatch-less composite kernel.
Here are the main changes:

- `Loss.cpp`
    - Removed `soft_margin_loss` function definitions
- `composite/soft_margin_loss.h`
    - Add the templated functions
- `native_functions.yaml`
    - Add the `composite` key to `soft_margin_loss`
- `generated_ops.cpp`
    - Replaced `at::native::soft_margin_loss` with `at::cpu::soft_margin_loss`

[ghstack-poisoned]

ezyang commented:
sure, I guess that is fine, but xla probably should also fix their kernel
